Add a simple command that knows how to download the dcos client.

brendandburns opened the pull request:
@rgardler @ahowe

This is my first PR to the Azure CLI, please advise if I'm doing it wrong...

tjprescott commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py:
```diff
@@ -59,6 +61,33 @@ def dcos_browse(name, resource_group_name):
+        install_location = '/usr/local/bin/dcos'
+        file_url = 'https://downloads.dcos.io/binaries/cli/darwin/x86-64/dcos-1.8/dcos'
+    else:
+        logger.error('unknown system: %s' % system)
```
This should probably raise a CLIError. Otherwise it will attempt to install to "".

brendandburns replied:
done.

tjprescott commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py:
```diff
@@ -59,6 +61,33 @@ def dcos_browse(name, resource_group_name):
+        logger.error('unknown system: %s' % system)
+
+    user_location = input('dcos binary install location [%s]:' % install_location)
+    if len(user_location) == 0:
```
if not len(user_location): ?

derekbekoe commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py:
```diff
@@ -59,6 +61,33 @@ def dcos_browse(name, resource_group_name):
 
     return
 
+def dcos_install():
```
`install_location` should be an optional argument to this function so it can be specified on the command line with `acs dcos install --install-location ...`

i.e.
```
def dcos_install(install_location=None):
   ...
```

derekbekoe replied:
You can the acs/_params.py file to define what install_location should be.
This means that when this method is called, the right install_location will be passed in.

So for example, you can remove the install locations that have been defined here and add the following lines to the `acs/_params.py` file:
```
import platform

def _get_default_install_location():
    system = platform.system()
    if system == 'Windows':
        return 'C:\Program Files\dcos.exe'
    if system == 'Linux':
        return '/usr/local/bin/dcos'
    if system == 'Darwin':
        return '/usr/local/bin/dcos'
    else:
        return None

register_cli_argument('acs dcos install', 'install_location', options_list=('--install-location',),
                      default=_get_default_install_location())
```

This means when the user runs `az acs dcos install -h`, they'll see:
```
Arguments
    --install-location                                        : Default: /usr/local/bin/dcos.
```
with the default being computed automatically.

brendandburns replied:
done.

derekbekoe commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py:
```diff
@@ -59,6 +61,33 @@ def dcos_browse(name, resource_group_name):
+    else:
+        logger.error('unknown system: %s' % system)
+
+    user_location = input('dcos binary install location [%s]:' % install_location)
```
Our convention is to not take user input interactively like this.
It makes it difficult to set the install_location with a script.
I'd suggest removing this and adding an optional argument `install_location` like I've described above.

brendandburns replied:
done.

derekbekoe commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py:
```diff
@@ -59,6 +61,33 @@ def dcos_browse(name, resource_group_name):
+        user_location = install_location
+
+    logger.info("Downloading client to %s\n" % user_location)
+    urllib.request.urlretrieve(file_url, user_location)
```
Should wrap this in a try/except as it can raise an `IOError` if the connection cannot be made (https://docs.python.org/2/library/urllib.html#urllib.urlopen).

e.g.
```
try:
    urllib.request.urlretrieve(file_url, user_location)
except IOError:
    raise CLIError('Connection error attempting to download client')
```

brendandburns replied:
done.

derekbekoe commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py:
```diff
@@ -59,6 +61,33 @@ def dcos_browse(name, resource_group_name):
+    if len(user_location) == 0:
+        user_location = install_location
+
+    logger.info("Downloading client to %s\n" % user_location)
```
Shouldn't need the '\n' in the log message

brendandburns replied:
done.

derekbekoe commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/generated.py:
```diff
@@ -8,4 +8,8 @@
+    dcos_install)
+
 cli_command('acs dcos browse', dcos_browse)
+cli_command('acs dcos install', dcos_install)
```
Missing new line that pylint was complaining about.

brendandburns replied:
done.

derekbekoe commented:
@brendandburns As an FYI, to prevent the 'cla-required' label from being applied, visit aka.ms/azuregithub to link your GitHub account.

derekbekoe commented:
Also, why is the command `acs dcos install` when it actually downloads the client only?

brendandburns commented:
@derekbekoe thanks for the fast review.  addressed all coments, pylint is clean now.

Changed to `acs dcos install-cli`

Please take another look.

Thanks!
--brendan

derekbekoe commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py:
```diff
@@ -3,10 +3,26 @@
+    elif system == 'Linux' or system == 'Darwin':
+        install_location = '/usr/local/bin/{}'.format(exeName)
+    else:
+        raise CLIError('Proxy server ({}) does not exist on the cluster.'.format(system))
```
Here, should just set `install_location=None`.
It would be too soon to raise a CLIError at this stage as it'll cause the 'acs' module to not load.

Then, in the `dcos_install_cli()` method, you'd have a check for:
```
def dcos_install_cli(install_location=None):
    if not install_location:
        raise CLIError("No install location specified and it could not be determined from the current platform '{}'".format(system))
```

brendandburns replied:
done.

derekbekoe commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py:
```diff
@@ -3,10 +3,26 @@
 from azure.cli.core.commands.parameters import (
     name_type,
     resource_group_name_type)
+from azure.cli.core._util import CLIError
```
After making my other suggested change, this import would be unused so can be removed.

brendandburns replied:
done.

derekbekoe commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py:
```diff
@@ -3,8 +3,10 @@
+import platform
 import random
 import string
+import urllib.request
```
The CI build is failing because of this.
`urllib.request` is defined in Python 3 but not Python 2.

This should fix this:
`from six.moves.urllib.request import urlretrieve`
from: https://pythonhosted.org/six/#module-six.moves.urllib.request

brendandburns replied:
done.

yugangw-msft commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py:
```diff
@@ -3,10 +3,26 @@
+def _get_default_install_location(exeName):
+    system = platform.system()
+    if system == 'Windows':
+        install_location = 'C:\\Program Files\\{}.exe'.format(exeName)
```
please use `os.environ['ProgramFiles']`

johanste replied:
x64 vs x86? And please verify behavior with both 64bit/32bit python on windows...

derekbekoe replied:
Also, `os.environ.get('ProgramFiles')` instead as then you don't get a KeyError from the dict and can check for None on the result and not set install location if you get none back from the call to os.environ.

brendandburns replied:
done.

yugangw-msft commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py:
```diff
@@ -59,6 +60,27 @@ def dcos_browse(name, resource_group_name):
+    file_url = ''
+    system = platform.system()
+    if system == 'Windows':
+        file_url = 'https://downloads.dcos.io/binaries/cli/windows/x86-64/dcos-1.8/dcos.exe'
```
Suggest to add a command help to call out the version of `1.8`, or expose the version as parameter people can pick; otherwise people will be surprised that we don't install newer stable one after some period.

brendandburns replied:
done.

brendandburns commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py:
```diff
@@ -3,10 +3,26 @@
     resource_group_name_type)
+from azure.cli.core._util import CLIError
+
+def _get_default_install_location(exeName):
```
done.

brendandburns commented:
Comments addressed, please re-check.

Thanks!
--brendan

yugangw-msft commented:
@brendandburns, to merge the PR, you will need to suppress the pylint error of "E:  9, 0: Unable to import 'six.moves.urllib.request' (import-error)". I assume you favor this lib over the `requests` package for the simplicity reason

brendandburns commented:
@yugangw-msft do I need to worry about adding that to a dependency list somewhere?

THanks
--brendan

yugangw-msft commented:
@brendandburns, `six` is already brought in through the common dependency of `azure-cli-core`. Still, I think you should update acs command package's `setup.py` with this dependency. The pylint error might still surface for you to suppress.
//cc:@derekbekoe

derekbekoe commented:
@brendandburns: As @yugangw-msft mentioned, add ` #pylint: disable=import-error` after the `six` import that's failing the pylint check.
Also, add `six` to the `DEPENDENCIES` section of the setup.py file that Yugang referenced.

brendandburns commented:
Done, please take another look.

Thanks!
--brendan

yugangw-msft commented on src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py:
```diff
@@ -3,8 +3,11 @@
 import random
 import string
+#pylint: disable=import-error
+from six.moves.urllib.request import urlretrieve
```
You probably should add the suppression at the same line ; otherwise the suppression will apply on all the imports in the rest of the file

brendandburns replied:
done.

brendandburns commented:
Comments addressed, please take another look.

Thanks!
--brendan
